salt/client/ssh/__init__.py

```diff
@@ -537,37 +549,38 @@ def handle_routine(self, que, opts, host, target, mine=False):
         Run the routine in a "Thread", put a dict on the queue
         """
         opts = copy.deepcopy(opts)
-        single = Single(
-            opts,
-            opts["argv"],
-            host,
-            mods=self.mods,
-            fsclient=self.fsclient,
-            thin=self.thin,
-            mine=mine,
-            **target
-        )
-        ret = {"id": single.id}
-        logging._acquireLock()
-        stdout, stderr, retcode = single.run()
-        logging._releaseLock()
-        # This job is done, yield
-        try:
-            data = salt.utils.json.find_json(stdout)
-            if len(data) < 2 and "local" in data:
-                ret["ret"] = data["local"]
-            else:
+        ret = {}
+        fake_lock = FakeLock()
+        with patch('importlib._bootstrap._get_module_lock', return_value=fake_lock):
```

Comment:
This seems really hacky, do we really need that? I see you removed the `_acquireLock` & `_releaseLock` calls below, so I assume it's needed for something else. What codepath needs this mocked?

Reply:
yes, it's hacky, but it's the only way I found to solve importlib deadlocks (no any side effects found in 24 hours stress test), it's not about `_acquireLock` and `_releaseLock` for `logging`. `importlib._bootstrap._get_module_lock` is returning individual lock objects for each module and in multiprocessing case it also leads to deadlocks, there are some changes for Python 3.8, but I tested salt in Tumbleweed with Python 3.8 and it seems that deadlocks with `importlib` are still there in Python 3.8

Reply:
>  importlib._bootstrap._get_module_lock is returning individual lock objects for each module and in multiprocessing case it also leads to deadlocks

Can you elaborate on this part? I don't understand the cause of the problem yet. I hope we find a cleaner solution than monkey patching part of the standard library in the code we ship to our customers.

Reply:
Here is the example of traceback for salt-api child process get stuck on `acquire`:
```
Traceback (most recent call first):
  File "<frozen importlib._bootstrap>", line 98, in acquire
  File "<frozen importlib._bootstrap>", line 149, in __enter__
  File "<frozen importlib._bootstrap>", line 586, in _exec
  File "<frozen importlib._bootstrap>", line 262, in _load_module_shim
  File "<frozen importlib._bootstrap_external>", line 852, in load_module
  File "<frozen importlib._bootstrap_external>", line 1027, in load_module
  File "<frozen importlib._bootstrap_external>", line 1034, in _check_name_wrapper
  File "/usr/lib/python3.8/site-packages/salt/loader.py", line 4779, in _load_module
  File "/usr/lib/python3.8/site-packages/salt/loader.py", line 1926, in _inner_load
    if self._load_module(name) and key in self._dict:
  File "/usr/lib/python3.8/site-packages/salt/loader.py", line 2193, in _load
  File "/usr/lib/python3.8/site-packages/salt/utils/lazy.py", line 99, in __getitem__
...
```
It could fail on loading different modules. I have not found any common pattern for it. You can take a look on the sources for lock object used in importlib and it seems it's not intended to be used with multiprocessing. And the bad thing that it's happening in frozen module.

Reply:
The example above was captured from Tumbleweed with Python 3.8 which still has the same issues with logging and importlib. There was a quite strange fix for importlib which most probably did not help at all: https://github.com/python/cpython/pull/17518